### PR TITLE
[REVIEW]Add support and tests  for cuML and XGBoost

### PR DESCRIPTION
**Pull request by VibhuJawa**
## Add support and tests for cuML and XGBoost 
This PR  addresses https://github.com/dask-contrib/dask-sql/issues/309

TODO:
- [x] Ensure efficient path for non distributed models and allow single GPU `xgboost` to succeed 
        
       For more context around this see issue: rapidsai/cuml#4406 .

       We were previously training on client which is:
         a. Very inefficient and possibly problematic in multi-node clusters and heterogeneous setup.
         b. Training non distributed xgboost models on dask collections is not supported.

- [x] Test for singe GPU `cuml` model
- [x] Test for multi gpu `cuml` model
- [x] Test for singe gpu `xgboost` model
- [x] Test for multi gpu `xgboost` model
- [x] Update GPU-CI environment with the right libraries
See PR: https://github.com/rapidsai/dask-build-environment/pull/16/files 

### Follow Up work to enable predict with multi GPU cuML models:
- [x] Triage single gpu `cuml` model failure 
See issue: https://github.com/dask-contrib/dask-sql/issues/332

### CODE REVIEW

File: tests/integration/test_model.py
Comment (VibhuJawa) [inline tests/integration/test_model.py:94]: Error Trace: 

```python

distributed.worker - WARNING - Compute Failed
Function:  _get_model_attr
args:      (LinearRegressionMG(), '__deepcopy__')
kwargs:    {}
Exception: 'AttributeError("Attribute __deepcopy__ does not exist on model <class \'cuml.linear_model.linear_regression_mg.LinearRegressionMG\'>")'
```

Probably due to creating a duplicate context at [line](https://github.com/dask-contrib/dask-sql/blob/ebdf4d5e5e5d622bfda881129620a4216f16e453/dask_sql/physical/rel/custom/predict.py#L91-L92) here .

**Comment by VibhuJawa** [comment]
CC: @charlesbluca , How do i go about adding `xgboost` and `cuML` to the `gpu-CI` we have setup ?

**Comment by charlesbluca** [comment]
You can open a PR adding these packages to the environment created in this dockerfile:

https://github.com/rapidsai/dask-build-environment/blob/main/dask_sql.Dockerfile

File: tests/integration/test_model.py
Comment (VibhuJawa) [reply tests/integration/test_model.py:94]: Raised https://github.com/dask-contrib/dask-sql/issues/332 to track this issue.

**Comment by VibhuJawa** [comment]
@charlesbluca , This is now ready for review. 

This PR is dependent on https://github.com/rapidsai/dask-build-environment/pull/16/files

File: dask_sql/physical/rel/custom/create_model.py
Comment (VibhuJawa) [inline dask_sql/physical/rel/custom/create_model.py:183]: For more context around this see issue: https://github.com/rapidsai/cuml/issues/4406 .

We were previously training on client which is:

a. Very inefficient and possibly problematic in multi-node clusters and heterogeneous setup. 
b. Training non distributed xgboost models on dask collections is not supported.

File: tests/integration/test_model.py
Comment (charlesbluca) [inline tests/integration/test_model.py:50]: Any reason we opt for `map_partitions` here over `dask_cudf.from_dask_dataframe`?

File: tests/integration/test_model.py
Comment (charlesbluca) [inline tests/integration/test_model.py:73]: We should mark these tests for GPU with `@pytest.mark.gpu`

**Review by charlesbluca — COMMENTED** [review]
Thanks for the work here @VibhuJawa 😄 a couple comments, I'm not the most knowledgeable on Dask ML stuff so would feel good getting a second review on this:

File: tests/integration/test_model.py
Comment (VibhuJawa) [reply tests/integration/test_model.py:73]: Thanks.  Let me mark those. 

Any place i can read up on how `@pytest.mark.gpu` works ?

File: tests/integration/test_model.py
Comment (VibhuJawa) [reply tests/integration/test_model.py:50]: No reason . Happy to switch it.

File: tests/integration/test_model.py
Comment (charlesbluca) [reply tests/integration/test_model.py:73]: Sure! This page generally goes over how classes/modules can be marked, and how to select/deselect them in pytest runs:

https://docs.pytest.org/en/6.2.x/example/markers.html

One thing to note is that dask-sql skips GPU tests by default unless `--rungpu` is specified:

https://github.com/dask-contrib/dask-sql/blob/96524ee41fd004a92330c32c3be35b5201e6b1c2/conftest.py#L10-L12

That shouldn't conflict with anything changes you've made here, but you should be safe to remove the `pytest.importorskip` lines in individual tests and replace them with a single `try..except ImportError` at the top of the file.

Comment (VibhuJawa) [reply tests/integration/test_model.py:85]: Resolved with commithttps://github.com/dask-contrib/dask-sql/pull/330/commits/f583c14e07027599b91a7fcc7e9d6d705e9853ab

**Review by ChrisJar — APPROVED** [review]
LGTM!

**Review by charlesbluca — COMMENTED** [review]
Missed this. but generally things look good here! Happy to merge this once cuML  / XGBoost are successfully added to gpuCI and tests are passing

File: tests/integration/fixtures.py
Comment (VibhuJawa) [reply tests/integration/fixtures.py:273]: No reason. Just felt the right thing to do in case when we  want to use an external scheduler later on .  `Dask-sql` has some separate tests with external cpu scheduler.   [Link code](https://github.com/dask-contrib/dask-sql/blob/61a20b78877d96e395b359e11c9a6a9c5c554cdb/.github/workflows/test.yml#L166-L179)

Then `gpu_client` can just be used accordingly.

**Review by charlesbluca — APPROVED** [review]

**Comment by VibhuJawa** [comment]
@GPUtester  rerun tests .

File: tests/integration/fixtures.py
Comment (charlesbluca) [inline tests/integration/fixtures.py:259]:
```suggestion
@pytest.fixture()
def gpu_cluster():
```

File: tests/integration/fixtures.py
Comment (charlesbluca) [inline tests/integration/fixtures.py:269]:
```suggestion
@pytest.fixture()
def gpu_client(gpu_cluster):
```

**Review by charlesbluca — CHANGES_REQUESTED** [review]
Think we need to mark these as fixtures to expose them to tests?

**Comment by charlesbluca** [comment]
rerun tests

**Comment by charlesbluca** [comment]
rerun tests

**Review by charlesbluca — APPROVED** [review]

**Pull request merged**
